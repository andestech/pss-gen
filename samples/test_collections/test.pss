/*!
@weakgroup	test
@{
@file
@author         Luther Lee
@data           2022/10/11
@brief          Test declaration, assignment, and index of collection.
@version
Portable Test and Stimulus Standard Version 2.0 <br>
> Section 8.8 : Collections <br>

### Test Command

    make test


@bug		PSSGenVisitor.visitMap_type(), Ln314: Syntax is not yet supported: 'map<string,int>'
@bug		PSSGenVisitor.visitSet_type(), Ln320: Syntax is not yet supported: 'set<int>'
@bug		PSS-ERROR: [LISTINDEX] '<pss_top>.<root_a>.<test>.my_list' index '1' is out of range
*/
component pss_top {
	action test {
		// Test declaration
		int array1[3];
		int array2[3];
		array<int, 3> my_array;
		list<int> my_list;
		map<string, int> my_map;
		map<string, string> my_map_string;
		map<int, string> my_map_int;
		// set<int> my_set;		// Fixme: PSSGenVisitor.visitSet_type(), Ln320: Syntax is not yet supported: 'set<int>'

		// Test assignment
		exec post_solve {
			array1[0] = 3;
			array1[1] = 2;
			array1[2] = 1;
			array2 = {13,12,11};
			my_array = {23, 22, 21};
			my_list = {33, 32, 31};
			my_map = {"a":43, "b":50, "c":41};
            my_map_string = {"x":"xxx", "y":"yyy", "z":"zzzz" };
            my_map_int = {10:"10", 20:"20", 30:"30" };
			// my_set = {53, 52, 51};
		}

		// Test index
		activity {
			do printf with {val == array1[1];};
			do printf with {val == array2[1];};
			do printf with {val == my_array[1];};
			do printf with {val == my_list[1];};
			do printf with {val == my_map["b"];};
			do print_string with {sss == my_map_string["x"];};
			do print_string with {sss == my_map_int[10];};
			// do printf with {val == my_set[1];};
		}
	}

	action printf {
		rand int val;
		exec body ASM =
		"""val = {{val}}""";
	}

    action print_string {
        rand string sss;
        exec body ASM =
        """sss = {{sss}}""";
    }

	action root_a {
		activity {
			do test;
		}
	}
}

/**@}*/
