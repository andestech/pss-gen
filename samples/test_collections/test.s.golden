val = 2
val = 12
val = 22
